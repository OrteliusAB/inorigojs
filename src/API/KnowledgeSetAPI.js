import axios from "axios"

/*
 * Knowledge Set API class.
 * This class contains APIs for interacting with knowledge sets.
 */
export class KnowledgeSetAPI {
	/**
	 * Knowledge Set API
	 * @param {InorigoAPI} parentAPI - The parent API class.
	 */
	constructor(parentAPI) {
		this.parentAPI = parentAPI
	}

	/**
	 * Retrieves meta data for a given knowledge set.
	 * @param {string} uuid - UUID of the knowledge set
	 * @return {object} - Response
	 */
	getMetaData(uuid) {
		return axios.post(`${this.parentAPI.BASE_URL_API}knowledgeset/${uuid}?metadata=true&page=1&pagesize=0`, {}, this.parentAPI.DEFAULTCONFIG)
	}

	/**
	 * Executes a given knowledge set and retrieves the response in a flat format.
	 * @param {string} uuid - UUID of the knowledge set
	 * @param {boolean} isDistinct - Should rows be distinct?
	 * @param {number} page - What page do you want to retrieve?
	 * @param {number} pagesize - How many rows should be in one page?
	 * @param {object} parameters - Extra parameters for the knowledge set provided as an object literal
	 * @param {boolean} allowCache - Allow cached data?
	 * @return {object} - Response
	 */
	getResult(uuid, isDistinct, page, pagesize, parameters, allowCache) {
		const uriParams = {
			metadata: true,
			distinct: isDistinct,
			page,
			pagesize
		}
		return axios.post(
			`${this.parentAPI.BASE_URL_API}knowledgeset/${uuid}${allowCache ? "/cache/read" : ""}${this.parentAPI._buildURIParams(uriParams)}`,
			parameters !== undefined && parameters !== null ? { parameters } : {},
			this.parentAPI.DEFAULTCONFIG
		)
	}

	/**
	 * Executes a free text search in a given knowledge set.
	 * @param {string} uuid - UUID of the knowledge set
	 * @param {string} text - Text string to search for
	 * @param {boolean} fuzzy - Is it a fuzzy search?
	 * @param {boolean} metaData - Include meta data?
	 * @param {boolean} compactLeafs - Should leafs be compact when searching?
	 * @param {boolean} allowCache - Allow cached data?
	 * @param {boolean} searchIDs - Should IDs be searched?
	 * @param {boolean} includedColumns - Columns of the knowledge set to be exclusively included
	 * @param {boolean} excludedColumns - Columns of th knowledge set to be exclusively excluded
	 * @return {object} - Response
	 */
	searchResult(uuid, text, fuzzy, metaData, compactLeafs, allowCache, searchIDs, includedColumns, excludedColumns) {
		const uriParams = {
			text,
			fuzzy,
			metadata: metaData,
			compactleafs: compactLeafs,
			allowCache,
			searchIDs,
			includedColumns,
			excludedColumns
		}
		return axios.get(
			`${this.parentAPI.BASE_URL_API}knowledgeset/${uuid}/tree/search/text${this.parentAPI._buildURIParams(uriParams)}`,
			this.parentAPI.DEFAULTCONFIG
		)
	}

	/**
	 * Executes a given knowledge set and retrieves the response in a tree format.
	 * @param {string} uuid - UUID of the knowledge set
	 * @param {boolean} metaData - Should meta data be included?
	 * @param {number} compactLeafs - Should leafs be made compact?
	 * @param {object} parameters - Extra parameters for the knowledge set provided as an object literal
	 * @param {boolean} allowCache - Allow cached data?
	 * @return {object} - Response
	 */
	getTreeResult(uuid, metaData, compactLeafs, parameters, allowCache) {
		const uriParams = {
			metadata: metaData,
			compactleafs: compactLeafs,
			allowCache
		}
		return axios.post(
			`${this.parentAPI.BASE_URL_API}knowledgeset/tree/${uuid}${this.parentAPI._buildURIParams(uriParams)}`,
			parameters !== undefined && parameters !== null ? { parameters } : {},
			this.parentAPI.DEFAULTCONFIG
		)
	}

	/**
	 * Executes a free text search in a given knowledge set in tree format.
	 * @param {string} uuid - UUID of the knowledge set
	 * @param {string} text - Text string to search for
	 * @param {boolean} fuzzy - Is it a fuzzy search?
	 * @param {boolean} metaData - Include meta data?
	 * @param {boolean} compactLeafs - Should leafs be compact when searching?
	 * @param {boolean} allowCache - Allow cached data?
	 * @param {boolean} searchIDs - Should IDs be searched?
	 * @param {boolean} includedColumns - Columns of the knowledge set to be exclusively included
	 * @param {boolean} excludedColumns - Columns of th knowledge set to be exclusively excluded
	 * @return {object} - Response
	 */
	searchTreeResult(uuid, text, fuzzy, metaData, compactLeafs, allowCache, searchIDs, includedColumns, excludedColumns) {
		const uriParams = {
			text,
			fuzzy,
			metadata: metaData,
			compactleafs: compactLeafs,
			allowCache,
			searchIDs,
			includedColumns,
			excludedColumns
		}
		return axios.get(
			`${this.parentAPI.BASE_URL_API}knowledgeset/${uuid}/tree/search/text${this.parentAPI._buildURIParams(uriParams)}`,
			this.parentAPI.DEFAULTCONFIG
		)
	}

	/**
	 * Get the definition of specified knowledgeset
	 * @param {string} uuid
	 * @returns {object} - Response
	 */
	getDefinition(uuid) {
		return axios.get(`${this.parentAPI.BASE_URL_API}knowledgeset/${uuid}/definition`, this.parentAPI.DEFAULTCONFIG)
	}

	/**
	 * Retrieves all knowledgesets found in Inorigo.
	 * @return {object} - Response
	 */
	getAvailable() {
		return axios.get(`${this.parentAPI.BASE_URL_API}knowledgeset/list`, this.parentAPI.DEFAULTCONFIG)
	}

	/**
	 * Executes a SQL query on top of a knowledge set result.
	 * @param {string} uuid - UUID of the knowledge set
	 * @param {string} query - SQL Query (MSSQL)
	 * @param {boolean} metaData - Include meta data?
	 * @param {string} context - Conext ID in Inorigo
	 * @param {object} parameters - Extra parameters for the knowledge set provided as an object literal
	 * @param {boolean} allowCache - Allow cached data?
	 * @return {object} - Response
	 */
	query(uuid, query, metadata, context, parameters, allowCache) {
		const uriParams = {
			sql: query,
			metadata,
			context,
			allowCache
		}
		return axios.post(
			`${this.parentAPI.BASE_URL_API}knowledgeset/${uuid}/query${this.parentAPI._buildURIParams(uriParams)}`,
			parameters !== undefined && parameters !== null ? { parameters } : {},
			this.parentAPI.DEFAULTCONFIG
		)
	}

	/**
	 * Counts the rows inside of a knowledge set.
	 * @param {string} uuid - UUID of the knowledge set
	 * @param {boolean} isDistinct- Should the rows be distinct?
	 * @return {object} - Response
	 */
	countRows(uuid, isDistinct) {
		return axios.post(
			`${this.parentAPI.BASE_URL_API}knowledgeset/${uuid}/count${this.parentAPI._buildURIParams({
				distinct: isDistinct
			})}`,
			{},
			this.parentAPI.DEFAULTCONFIG
		)
	}

	/**
	 * Retrieves all knowledge set scheduling information.
	 * @return {object} - Response
	 */
	getSchedulingStatus() {
		return axios.get(`${this.parentAPI.BASE_URL_API}knowledgeset/scheduling/status`, this.parentAPI.DEFAULTCONFIG)
	}

	/**
	 * Export knowledgeset to a file, as .CSV or Excel format, filename will be same as knowledgeset name.
	 * @param {string} uuid
	 * @param {boolean} metaData
	 * @param {string} context
	 * @param {boolean} isDistinct
	 * @param {number} page
	 * @param {number} pagesize
	 * @param {boolean} compactpaths
	 * @param {boolean} allowCache
	 * @param {boolean} replaceidbypresentation
	 * @param {boolean} excelFormat
	 * @returns {object} - Response
	 */
	exportToFile(uuid, metaData, context, isDistinct, page, pagesize, compactpaths, allowCache, replaceidbypresentation, excelFormat = false) {
		const uriParams = {
			metadata: metaData,
			context,
			distinct: isDistinct,
			page,
			pagesize,
			compactpaths,
			allowCache,
			replaceidbypresentation
		}
		// default text/csv
		this.parentAPI.setDefaultRequestHeader("Accept", "text/csv")
		if (excelFormat) {
			this.parentAPI.setDefaultRequestHeader("Accept", "application/vnd.ms-excel")
		}
		return axios.get(`${this.parentAPI.BASE_URL_API}knowledgeset/file/${uuid}${this.parentAPI._buildURIParams(uriParams)}`, this.parentAPI.DEFAULTCONFIG)
	}

	/**
	 * Execute a Knowledge Set. Result as Json Objects.
	 * @param {string} uuid
	 * @param {boolean} metaData
	 * @param {string} context
	 * @param {boolean} isDistinct
	 * @param {number} page
	 * @param {number} pagesize
	 * @param {boolean} compactpaths
	 * @param {boolean} allowCache
	 * @param {boolean} replaceidbypresentation
	 * @returns {object} - Response
	 */
	getResultAsObjects(uuid, metaData, context, isDistinct, page, pagesize, compactpaths, allowCache, replaceidbypresentation) {
		const uriParams = {
			metadata: metaData,
			context,
			distinct: isDistinct,
			page,
			pagesize,
			compactpaths,
			allowCache,
			replaceidbypresentation
		}
		// this.parentAPI.setDefaultRequestHeader("Accept", "application/json")
		// this.parentAPI.setDefaultRequestHeader("Content-Type", "application/json")
		return axios.post(
			`${this.parentAPI.BASE_URL_API}knowledgeset/objects/${uuid}${this.parentAPI._buildURIParams(uriParams)}`,
			{},
			this.parentAPI.DEFAULTCONFIG
		)
	}

	/**
	 * Search Knowledge Set for text occurrences. Result as Json Objects.
	 * @param {string} uuid - The uuid of the Knowledge Set to search
	 * @param {string} text - The text to search for
	 * @param {boolean} fuzzy - Optional parameter. Enable Fuzzy Pattern Match? Default is false.
	 * @param {boolean} metaData - Optional boolean parameter. If true, the response will include metadata about the Knowledge Set
	 * @param {number} page - The page index
	 * @param {number} pagesize - Number of items per page
	 * @param {boolean} compactpaths - Compact Paths
	 * @param {boolean} allowCache - Optional parameter. Allow data to be read from cache?
	 * @param {boolean} searchIDs - Optional parameter. Should ID columns be searched?
	 * @param {Array} includedColumns - Optional parameter. Columns to be included in search
	 * @param {Array} excludedColumns - Optional parameter. Columns to be excluded from search
	 * @param {boolean} replaceidbypresentation - Optional parameter. Replace all ids by presentations?
	 * @returns
	 */
	searchResultAsObjects(
		uuid,
		text,
		fuzzy,
		metaData,
		page,
		pagesize,
		compactpaths,
		allowCache,
		searchIDs,
		includedColumns,
		excludedColumns,
		replaceidbypresentation
	) {
		const uriParams = {
			text,
			fuzzy,
			metadata: metaData,
			page,
			pagesize,
			compactpaths,
			allowCache,
			searchIDs,
			includedColumns,
			excludedColumns,
			replaceidbypresentation
		}
		return axios.get(
			`${this.parentAPI.BASE_URL_API}knowledgeset/${uuid}/objects/search/text${this.parentAPI._buildURIParams(uriParams)}`,
			this.parentAPI.DEFAULTCONFIG
		)
	}
}
