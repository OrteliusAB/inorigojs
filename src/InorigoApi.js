import axios from "axios";
import * as https from 'https';

/*
 * This is the main API communication class.
 * This class contains functions that return communication instances of each respective API of Inorigo, but with one shared API class instance.
 * The class uses Axios under the hood to do the heavy lifting.
*/
export default class InorigoAPI {

  constructor(url, { rejectUnauthorized, authorization, apiEndpoint }) {
    this.IS_SECURE = true
    this.BASE_URL = url;
    this.BASE_URL_API = url += "services/api/v1/"
    this.DEFAULTCONFIG = {
      headers: {
        "Content-Type": "application/json;charset=utf-8",
        "Accept": "application/json",
        "Access-Control-Max-Age": 1728000
      },
      data: {},
      withCredentials: true
    }

    if (apiEndpoint) {
      this.BASE_URL_API = this.BASE_URL + apiEndpoint
    }

    if (rejectUnauthorized !== undefined) {
      this.IS_SECURE = rejectUnauthorized
      const agent = new https.Agent({
        rejectUnauthorized: rejectUnauthorized
      })
      this.DEFAULTCONFIG = {
        ...this.DEFAULTCONFIG,
        httpsAgent: agent
      }
    }

    if (authorization) {
      if (authorization.username && authorization.password) {
        this.DEFAULTCONFIG.headers.Authorization = "Basic " +
          this.genericBtoA(authorization.username + ":" + authorization.password)
      }
    }
  }

  /* Security, Global Settings */

  injectCookies(cookies) {
    this.DEFAULTCONFIG.headers.Cookie = cookies
  }

  /* Account, Session */

  login(username, password) {
    const CONFIG = {
      headers: {
        "Content-Type": "application/json;charset=utf-8",
        Accept: "application/json",
        Authorization: "Basic " + this.genericBtoA(username + ":" + password)
      },
      data: {},
      withCredentials: true
    }
    if (!this.IS_SECURE) {
      CONFIG.httpsAgent = this.DEFAULTCONFIG.httpsAgent
    }
    return axios.post(`${this.BASE_URL}services/open/login`, {}, CONFIG);
  }

  logout() {
    return axios.post(`${this.BASE_URL}services/open/logout`, {}, this.DEFAULTCONFIG);
  }

  getSession() {
    return axios.get(`${this.BASE_URL}services/open/session/user`, this.DEFAULTCONFIG);
  }

  /* API Shortcuts (Common inorigo helper functions) */

  shortcuts() {
    return {

      getClassifiedValuesList: (definitionUUID, isDeep) => {
        return new Promise((resolve, reject) => {
          this.entity().getSubClasses("AsDefinition", definitionUUID, isDeep).then(response => {
            const promises = response.data.entities.map(entity => {
              return this.entity().getEntity("AsDefinition", entity.id)
            })
            Promise.all(promises).then(result => {
              const options = result.map(genderOption => {
                return {
                  presentation: genderOption.data.presentation,
                  id: genderOption.data.id,
                  abstract: genderOption.data.attributeValues.filter(item => { return item.name === "abstract" })[0].value
                }
              })
              resolve(options)
            })
          })
        })
      }

    }
  }

  /* Verso (Application Builder) */

  versoRuntime() {
    return {

      refresh: (vrid) => {
        return axios.post(`${this.BASE_URL_API}application/runtime/${vrid}/refresh`, {}, this.DEFAULTCONFIG)
      },

      calculate: (vrid) => {
        return axios.post(`${this.BASE_URL_API}application/runtime/${vrid}/calculate`, {}, this.DEFAULTCONFIG)
      },

      clear: (vrid, where) => {
        return axios.post(`${this.BASE_URL_API}application/runtime/${vrid}/select/clear${this.buildURIParams({
          where: where,
        })}`,
          {}, this.DEFAULTCONFIG)
      },

      getValue: (vrid, where) => {
        return axios.get(`${this.BASE_URL_API}application/runtime/${vrid}/variable/value?key=${where}`, this.DEFAULTCONFIG)
      },

      selectOne: (vrid, where, what, isClearFirst) => {
        return axios.post(`${this.BASE_URL_API}application/runtime/${vrid}/select/one${this.buildURIParams({
          where: where,
          what: what,
          clear: isClearFirst
        })}`,
          {}, this.DEFAULTCONFIG)
      },

      selectMany: (vrid, selectionJSON) => {
        return axios.post(`${this.BASE_URL_API}application/runtime/${vrid}/select/many`, selectionJSON, this.DEFAULTCONFIG)
      },

      countAll: (vrid, where) => {
        return axios.get(`${this.BASE_URL_API}application/runtime/${vrid}/items/count/all?where=${where}`, this.DEFAULTCONFIG)
      },

      countSelected: (vrid, where) => {
        return axios.get(`${this.BASE_URL_API}application/runtime/${vrid}/items/count/selected?where=${where}`, this.DEFAULTCONFIG)
      },

      countExplicit: (vrid, where) => {
        return axios.get(`${this.BASE_URL_API}application/runtime/${vrid}/items/count/explicit?where=${where}`, this.DEFAULTCONFIG)
      },

      countImplicit: (vrid, where) => {
        return axios.get(`${this.BASE_URL_API}application/runtime/${vrid}/items/count/implicit?where=${where}`, this.DEFAULTCONFIG)
      },

      getScript: (vrid) => {
        return axios.get(`${this.BASE_URL_API}application/runtime/${vrid}/script`, this.DEFAULTCONFIG)
      },
    }
  }

  /* Knowledgeset (Flattened Structure) */

  knowledgeset() {
    return {

      getMetaData: uuid => {
        return axios.post(`${this.BASE_URL_API}knowledgeset/${uuid}?metadata=true&page=1&pagesize=0`, {}, this.DEFAULTCONFIG)
      },

      getResult: (uuid, isDistinct, page, pagesize, parameters, allowCache) => {
        let uriParams = {
          metadata: true,
          distinct: isDistinct,
          page: page,
          pagesize: pagesize
        }
        if (allowCache !== undefined && allowCache !== null) {
          uriParams.allowcache = "" + allowCache
        }
        return axios.post(`${this.BASE_URL_API}knowledgeset/${uuid}${this.buildURIParams(uriParams)}`,
          parameters !== undefined && parameters !== null ? { parameters: parameters } : {},
          this.DEFAULTCONFIG)
      },

      getTreeResult: (uuid, metaData, compactLeafs, parameters, allowCache) => {
        let uriParams = {
          metadata: metaData,
          compactleafs: compactLeafs,
        }
        if (allowCache !== undefined && allowCache !== null) {
          uriParams.allowcache = "" + allowCache
        }
        return axios.post(`${this.BASE_URL_API}knowledgeset/tree/${uuid}${this.buildURIParams(uriParams)}`,
          parameters !== undefined && parameters !== null ? { parameters: parameters } : {},
          this.DEFAULTCONFIG)
      },

      getAvailable: () => {
        return axios.get(`${this.BASE_URL_API}knowledgeset/list`, this.DEFAULTCONFIG)
      },

      query: (uuid, query) => {
        return axios.post(`${this.BASE_URL_API}knowledgeset/${uuid}/query?sql=${query}`, {}, this.DEFAULTCONFIG)
      },

      countRows: (uuid, isDistinct) => {
        return axios.post(`${this.BASE_URL_API}knowledgeset/${uuid}/count${this.buildURIParams({
          distinct: isDistinct
        })}`,
          {}, this.DEFAULTCONFIG)
      }

    }
  }

  /* Entity (Data Objects) */

  entity() {
    return {

      generateUUID: count => {
        return axios.get(`${this.BASE_URL_API}entity/generateid${this.buildURIParams({
          count: count
        })}`,
          this.DEFAULTCONFIG)
      },

      getEntity: (entityType, uuid) => {
        return axios.get(`${this.BASE_URL_API}entity/${entityType}/${uuid}/`, this.DEFAULTCONFIG)
      },

      getinstances: (entityType, uuid, informationType, page, pagesize) => {
        return axios.get(`${this.BASE_URL_API}entity/${entityType}/${uuid}/instances${this.buildURIParams({
          info: informationType,
          page: page,
          pagesize: pagesize
        })}`,
          this.DEFAULTCONFIG)
      },

      partners: (entityType, uuid, relationUuid, direction, isRecursive, isLeafsOnly, informationType, page, pagesize) => {
        return axios.get(`${this.BASE_URL_API}entity/${entityType}/${uuid}/partners${this.buildURIParams({
          relation: relationUuid,
          direction: direction,
          recursive: isRecursive,
          leafsonly: isLeafsOnly,
          info: informationType,
          page: page,
          pagesize: pagesize
        })}`,
          this.DEFAULTCONFIG)
      },

      getDefinitions: (entityType, uuid, isDeep, informationType, page, pagesize) => {
        return axios.get(`${this.BASE_URL_API}entity/${entityType}/${uuid}/definitions${this.buildURIParams({
          deep: isDeep,
          info: informationType,
          page: page,
          pagesize: pagesize
        })}`,
          this.DEFAULTCONFIG)
      },

      getSuperClasses: (entityType, uuid, isDeep, informationType, page, pagesize) => {
        return axios.get(`${this.BASE_URL_API}entity/${entityType}/${uuid}/superclasses?deep=${this.buildURIParams({
          deep: isDeep,
          info: informationType,
          page: page,
          pagesize: pagesize
        })}`,
          this.DEFAULTCONFIG)
      },

      getSubClasses: (entityType, uuid, isDeep, informationType, page, pagesize) => {
        return axios.get(`${this.BASE_URL_API}entity/${entityType}/${uuid}/subclasses${this.buildURIParams({
          deep: isDeep,
          info: informationType,
          page: page,
          pagesize: pagesize
        })}`,
          this.DEFAULTCONFIG)
      },

      getReferents: (entityType, uuid, page, pagesize) => {
        return axios.get(`${this.BASE_URL_API}entity/${entityType}/${uuid}/referents${this.buildURIParams({
          page: page,
          pagesize: pagesize
        })}`,
          this.DEFAULTCONFIG)
      },

      getRelations: (entityType, uuid, relationUuid, direction, page, pagesize) => {
        return axios.get(`${this.BASE_URL_API}entity/${entityType}/${uuid}/relations${this.buildURIParams({
          specifier: relationUuid,
          direction: direction,
          page: page,
          pagesize: pagesize
        })}`,
          this.DEFAULTCONFIG)
      },

      updateEntity: entityJSON => {
        return axios.put(`${this.BASE_URL_API}entity`, entityJSON, this.DEFAULTCONFIG)
      },

      createEntity: entityJSONArray => {
        return axios.post(`${this.BASE_URL_API}entity`, entityJSONArray, this.DEFAULTCONFIG)
      },
      transaction: transactionData => {
        return axios.post(`${this.BASE_URL_API}entity/commit/transaction`, transactionData, this.DEFAULTCONFIG)
      },

      deleteEntity: (entityType, uuid) => {
        return axios.delete(`${this.BASE_URL_API}entity/${entityType}/${uuid}`, this.DEFAULTCONFIG)
      },

      getSimplifiedInstances: (definitionUUID) => {
        return axios.get(`${this.BASE_URL_API}entity/AsDefinition/${definitionUUID}/instances?info=values`,
          this.DEFAULTCONFIG)
          .then(result => {
            const simplifiedJSON = result.data.entities.map(entity => {
              const newEntity = {
                id: entity.id,
                type: entity.dataType
              }
              newEntity.values = entity.attributeValues.reduce((simplifiedOilEntity, attributeValue) => {
                simplifiedOilEntity[attributeValue.name] = attributeValue.value
                return simplifiedOilEntity
              }, {})
              return newEntity
            })
            return new Promise((resolve, reject) => {
              resolve(simplifiedJSON)
            })
          })
      }

    }
  }

  /* Resource (Files) */

  resource() {
    return {

      getResource: (uuid) => {
        return axios.get(`${this.BASE_URL_API}resource/resource/${uuid}/`, this.DEFAULTCONFIG)
      },

      deleteResource: (uuid) => {
        return axios.delete(`${this.BASE_URL_API}resource/resource/${uuid}`, this.DEFAULTCONFIG)
      },

      createResource: resourceJSONArray => {
        return axios.post(`${this.BASE_URL_API}resource/resource`, resourceJSONArray, this.DEFAULTCONFIG)
      },

      updateResource: resourceJSONArray => {
        return axios.put(`${this.BASE_URL_API}resource/resource`, resourceJSONArray, this.DEFAULTCONFIG)
      }

    }
  }

  /* Utility Functions */

  buildURIParams(object) {
    let returnString = ""
    //Stringify removes the undefined values
    let tempObj = JSON.stringify(object)
    if (tempObj !== JSON.stringify({})) {
      returnString = "?"
    }
    tempObj = JSON.parse(tempObj)
    return (
      returnString +
      Object.keys(tempObj)
        .map(key => {
          return [key, tempObj[key]].map(encodeURIComponent).join("=")
        })
        .join("&")
    )
  }

  genericBtoA(b) {
    if (typeof process === 'object' && Object.prototype.toString.call(process) === '[object process]') {
      //If running in node, use Buffer
      return Buffer.from(b).toString("base64")
    }
    else {
      //If running in browser, use btoa()
      return btoa(b)
    }
  }

}
