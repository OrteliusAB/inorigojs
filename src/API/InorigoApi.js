import axios from "axios"
import { EntityAPI } from "./EntityAPI"
import { KnowledgeSetAPI } from "./KnowledgeSetAPI"
import { LegacyAPI } from "./LegacyAPI"
import { ResourceAPI } from "./ResourceAPI"
import { ShortcutAPI } from "./ShortcutAPI"
import { VersoRuntimeAPI } from "./VersoRuntimeAPI"

/*
 * This is the main API communication class.
 * This class contains functions that return communication instances of each respective API of Inorigo, but with one shared API class instance.
 */
export class InorigoAPI {
	/**
	 * Inorigo API
	 * @param {string} url - The URL of your inorigo instance.
	 * @param {object} options - Options
	 * @param {object} options.customHttpsAgent - A custom HTTPS agent that can used to, for example, surpress certificate related errors
	 * @param {object} options.authorization - Authorization parameters (username, password). If provided these will be stored and sent in every subsequent request
	 * @param {object} options.apiEndpoint - By default the latest version of the web API will typically be used. You can override this by providing your own endpoint.
	 */
	constructor(url, { customHttpsAgent, authorization, apiEndpoint }) {
		this.IS_SECURE = true
		this.BASE_URL = url
		this.BASE_URL_API = url += "services/api/v1/"
		this.DEFAULTCONFIG = {
			headers: {
				"Content-Type": "application/json;charset=utf-8",
				Accept: "application/json",
				"Access-Control-Max-Age": 1728000
			},
			data: {},
			withCredentials: true
		}

		if (apiEndpoint) {
			this.BASE_URL_API = this.BASE_URL + apiEndpoint
		}

		if (customHttpsAgent !== undefined) {
			this.IS_SECURE = false
			this.DEFAULTCONFIG = {
				...this.DEFAULTCONFIG,
				httpsAgent: customHttpsAgent
			}
		}

		if (authorization) {
			if (authorization.username && authorization.password) {
				this.DEFAULTCONFIG.headers.Authorization = "Basic " + this._genericBtoA(authorization.username + ":" + authorization.password)
			}
		}
	}

	/* Security, Global Settings */

	/**
	 * Can be used to inject cookies in all subsequent requests.
	 * @param {string} cookies - the cookie string
	 */
	injectCookies(cookies) {
		this.DEFAULTCONFIG.headers.Cookie = cookies
	}

	/* Account, Session */

	/**
	 * Send a login request to Inorigo.
	 * @param {string} username - Username to be used
	 * @param {string} password - Password to be used
	 */
	async login(username, password) {
		const CONFIG = {
			headers: {
				"Content-Type": "application/json;charset=utf-8",
				Accept: "application/json",
				Authorization: "Basic " + this._genericBtoA(username + ":" + password)
			},
			data: {},
			withCredentials: true
		}
		if (!this.IS_SECURE) {
			CONFIG.httpsAgent = this.DEFAULTCONFIG.httpsAgent
		}
		return axios.post(`${this.BASE_URL}services/open/login`, {}, CONFIG)
	}

	/**
	 * Send a logout request to Inorigo.
	 */
	async logout() {
		return axios.post(`${this.BASE_URL}services/open/logout`, {}, this.DEFAULTCONFIG)
	}

	/**
	 * Retrieves the current session from Inorigo.
	 */
	async getSession() {
		return axios.get(`${this.BASE_URL}services/open/session/user`, this.DEFAULTCONFIG)
	}

	/* API Shortcuts (Common inorigo helper functions) */

	/**
	 * Retrieves a shortcut API that can be used to abstract away some of the trickiness of the entity API.
	 * @return {ShortcutAPI} - The API
	 */
	getShortcutAPI() {
		return new ShortcutAPI(this)
	}

	/* DEPRECATED */
	shortcut() {
		return new ShortcutAPI(this)
	}

	/* Verso (Application Builder) */
	/**
	 * Retrieves an Application Builder API.
	 * @return {VersoRuntimeAPI} - The API
	 */
	getVersoRuntimeAPI() {
		return new VersoRuntimeAPI(this)
	}

	/* DEPRECATED */
	versoRuntime() {
		return new VersoRuntimeAPI(this)
	}

	/* Knowledge set */
	/**
	 * Retrieves a Knowledge Set API.
	 * @return {KnowledgeSetAPI} - The API
	 */
	getKnowledgesetAPI() {
		return new KnowledgeSetAPI(this)
	}

	/* DEPRECATED */
	knowledgeset() {
		return new KnowledgeSetAPI(this)
	}

	/* Entity (Data Objects) */
	/**
	 * Retrieves an Entity API.
	 * @return {EntityAPI} - The API
	 */
	getEntityAPI() {
		return new EntityAPI(this)
	}

	/* DEPRECATED */
	entity() {
		return new EntityAPI(this)
	}

	/* Resource (Files) */
	/**
	 * Retrieves a resource (file) API.
	 * @return {ResourceAPI} - The API
	 */
	getResourceAPI() {
		return new ResourceAPI(this)
	}

	/* DEPRECATED */
	resource() {
		return new ResourceAPI(this)
	}

	/* Legacy Endpoints (some of which are undocumented!) */
	/**
	 * Retrieves a legacy API.
	 * @return {LegacyAPI} - The API
	 */
	getLegacyAPI() {
		return new LegacyAPI(this)
	}

	/* DEPRECATED */
	legacy() {
		return new LegacyAPI(this)
	}

	/* Utility Functions */

	_buildURIParams(object) {
		let returnString = ""
		//Stringify removes the undefined values
		let tempObj = JSON.stringify(object)
		if (tempObj !== JSON.stringify({})) {
			returnString = "?"
		}
		tempObj = JSON.parse(tempObj)
		return (
			returnString +
			Object.keys(tempObj)
				.map(key => {
					if (Array.isArray(tempObj[key])) {
						let uriShard = ""
						tempObj[key].forEach(value => {
							uriShard = `${uriShard}&${key}=${value}`
						})
						return encodeURIComponent(uriShard.substr(1))
					}
					return [key, "" + tempObj[key]].map(encodeURIComponent).join("=")
				})
				.join("&")
		)
	}

	_genericBtoA(b) {
		if (typeof process === "object" && Object.prototype.toString.call(process) === "[object process]") {
			//If running in node, use Buffer
			return Buffer.from(b).toString("base64")
		} else {
			//If running in browser, use btoa()
			return btoa(b)
		}
	}
}
