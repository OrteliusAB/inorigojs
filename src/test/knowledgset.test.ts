import { assert, describe, expect, it } from "vitest"
import { InorigoAPI } from "../API/InorigoApi"
import { KnowledgeSetAPI } from "../API/KnowledgeSetAPI"
import { Utilities } from "../test-utils/utilities"

const utilities: Utilities = new Utilities()
const inorigoAPI: InorigoAPI = utilities.getInorigoAPI()
const knowledgesetAPI: KnowledgeSetAPI = inorigoAPI.getKnowledgesetAPI()

describe("Knowledgeset Module", () => {
	it("Utilities test, verify config read", () => {
		// utilities = new Utilities()
		assert.exists(utilities)
	})

	it("initialize IniorigoAPI to endpoint", () => {
		// console.log(inorigoAPI)
		assert.exists(inorigoAPI)
	})

	it.skip("check connection [/knowledgeset/list]", async () => {
		const list = await knowledgesetAPI.getAvailable()
		expect(list.status).equals(200)
	})

	it.skip("countRows(...) [/knowledgeset/{id}/count]", async () => {
		KSAPI = inorigoAPI.getKnowledgesetAPI()
		// const list = await KSAPI.countRows(uuid, isDistinct)
		// expect(list.status).equals(200)
	})

	it.skip("getResult(...) [/knowledgeset/{id}]", async () => {
		// const list = await KSAPI.countRows(uuid, isDistinct)
		// expect(list.status).equals(200)
	})

	it.skip("getResultAsObjects(...) [/knowledgeset/objects/{id}]", async () => {
		// const list = await KSAPI.countRows(uuid, isDistinct)
		// expect(list.status).equals(200)
	})

	it.skip("getResultAsObjects(...) [/knowledgeset/objects/{id}]", async () => {
		// const list = await KSAPI.countRows(uuid, isDistinct)
		// expect(list.status).equals(200)
	})

	it.skip("getTreeResult(...) [/knowledgeset/tree/{id}]", async () => {
		// const list = await KSAPI.countRows(uuid, isDistinct)
		// expect(list.status).equals(200)
	})

	it.skip("exportToFile(...) [/knowledgeset/file/{id}]", async () => {
		// const list = await KSAPI.countRows(uuid, isDistinct)
		// expect(list.status).equals(200)
	})

	it.skip("getDefinition(...) [/knowledgeset/{id}/definition]", async () => {
		// const list = await KSAPI.countRows(uuid, isDistinct)
		// expect(list.status).equals(200)
	})

	it.skip("getSchedulingStatus() [/knowledgeset/scheduling/status]", async () => {
		// const list = await KSAPI.countRows(uuid, isDistinct)
		// expect(list.status).equals(200)
	})

	it.skip("getAvailable() [/knowledgeset/list]", async () => {
		// const list = await KSAPI.countRows(uuid, isDistinct)
		// expect(list.status).equals(200)
	})

	it.skip("query(...)  [/knowledgeset/{id}/query]", async () => {
		// const list = await KSAPI.countRows(uuid, isDistinct)
		// expect(list.status).equals(200)
	})

	it.skip("getResult(...)  [/knowledgeset/{id}/cache/read]", async () => {
		// const list = await KSAPI.countRows(uuid, isDistinct)
		// expect(list.status).equals(200)
	})

	it.skip("searchResult(...) [/knowledgeset/{id}/search/text]", async () => {
		// const list = await KSAPI.countRows(uuid, isDistinct)
		// expect(list.status).equals(200)
	})

	it.skip("searchResultAsObjects(...) [/knowledgeset/{id}/objects/search/text]", async () => {
		// const list = await KSAPI.countRows(uuid, isDistinct)
		// expect(list.status).equals(200)
	})

	it.skip("searchTreeResult(...) [/knowledgeset/{id}/tree/search/text]", async () => {
		// const list = await KSAPI.countRows(uuid, isDistinct)
		// expect(list.status).equals(200)
	})
})
