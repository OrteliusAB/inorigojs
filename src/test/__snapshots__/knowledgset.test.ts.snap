// Vitest Snapshot v1

exports[`Knowledgeset Module > snapshot 1`] = `
{
  "foo": "bar",
}
`;
